kernel/src/schema/diff.rs

```diff
@@ -0,0 +1,1739 @@
+            .collect();
+        assert_eq!(array_updates.len(), 0);
+    }
+
+    #[test]
+    fn test_ancestor_filtering() {
+        // Test that when a parent struct is added, its children aren't reported separately
+        let current =
+            StructType::new_unchecked([create_field_with_id("id", DataType::LONG, false, 1)]);
+
+        let new = StructType::new_unchecked([
+            create_field_with_id("id", DataType::LONG, false, 1),
+            create_field_with_id(
+                "user",
+                DataType::try_struct_type([
+                    create_field_with_id("name", DataType::STRING, false, 3),
+                    create_field_with_id("email", DataType::STRING, true, 4),
+                    create_field_with_id(
+                        "address",
+                        DataType::try_struct_type([
+                            create_field_with_id("street", DataType::STRING, false, 6),
+                            create_field_with_id("city", DataType::STRING, false, 7),
+                        ])
+                        .unwrap(),
+                        true,
+                        5,
+                    ),
+                ])
+                .unwrap(),
+                false,
+                2,
+            ),
+        ]);
+
+        let diff = SchemaDiffArgs {
+            current: &current,
+            new: &new,
+        }
+        .compute_diff()
+        .unwrap();
+
+        // Only the top-level "user" field should be reported as added
+        assert_eq!(diff.added_fields.len(), 1);
+        assert_eq!(diff.added_fields[0].path, ColumnName::new(["user"]));
+
+        // No nested fields should be reported as added or updated
+        assert_eq!(diff.updated_fields.len(), 0);
+
+        // The filtered paths would have been: user.name, user.email, user.address, user.address.street, user.address.city
+    }
+
+    #[test]
+    fn test_ancestor_filtering_with_mixed_changes() {
+        let current = StructType::new_unchecked([
+            create_field_with_id("existing", DataType::STRING, false, 1),
+            create_field_with_id(
+                "existing_struct",
+                DataType::try_struct_type([create_field_with_id(
+                    "old_name",
+                    DataType::STRING,
+                    false,
+                    3,
+                )])
+                .unwrap(),
+                false,
+                2,
+            ),
+        ]);
+
+        let new = StructType::new_unchecked([
+            create_field_with_id("existing", DataType::STRING, true, 1), // Changed nullability
+            create_field_with_id(
+                "existing_struct",
+                DataType::try_struct_type([
+                    create_field_with_id("new_name", DataType::STRING, false, 3), // Renamed
+                ])
+                .unwrap(),
+                false,
+                2,
+            ),
```

Comment:
An important case to check is if both the ancestor and the child are changed, then both of them are included.

Example:

```suggestion
            create_field_with_id(
                "existing_struct",
                DataType::try_struct_type([
                    create_field_with_id("new_name", DataType::STRING, false, 3), // Renamed
                ])
                .unwrap(),
                true,    // Changed nullability
                2,
            ),
```

result should have both `existing_struct`, and `new_name`.